tools/ollama.go

```diff
@@ -120,3 +128,203 @@ func (*ollama) Install(logger *log.Logger, installPath string) (err error) {
 
 	return nil
 }
+
+var (
+	// OllamaURL holds the URL to the Ollama service.
+	OllamaURL = "http://127.0.0.1:11434"
+
+	// ollamaProcessStates holds Ollama URLs to states of Ollama service processes.
+	ollamaProcessStates = map[string]*ollamaProcessState{}
+	// ollamaProcessStatesLock holds the lock for accessing Ollama process states.
+	ollamaProcessStatesLock sync.Mutex
+)
+
+type ollamaProcessState struct {
+	// url holds the URL of the Ollama process.
+	url string
+	// connectionCount holds the amount of connections.
+	// When the counter hits `0`, the process will be removed.
+	connectionCount uint
+	// cleanup holds function to clean up the Ollama process.
+	// Can be `nil` in case the process was running already.
+	cleanup func()
+}
+
+// ollamaProcessStateAdd add a connection to the process states.
+func ollamaProcessStateAdd(url string, cleanup func()) {
+	state, ok := ollamaProcessStates[url]
+	if !ok {
+		state = &ollamaProcessState{
+			url: url,
+		}
+		ollamaProcessStates[url] = state
+	}
+	if cleanup != nil {
+		state.cleanup = cleanup // A process can be started afterwards as well, so make sure it is always cleaned up.
+	}
```

Comment:
corner case: if the state already existed and had a cleanup function, that function will be overwritten